Add Russound Binding

tmrobert8 opened the pull request:
Implementation of new Russound binding (RIO)

Signed-off-by: Tim Roberts troberts@bigfoot.com (github: tmrobert8)

tmrobert8 commented:
Kai,

Do you know what the issue is with this one (when I try to click on details - I get a 404 error)?  I can merge both of those commits into one - but I doubt that's the issue

Thanks,
Tim

tmrobert8 commented:
Well I merged the two commits and along the way noticed that I had a case issue (local was 1225-Russound but I had two remote ones "1225-russound" and "1225-Russound").  Removed the lowercase one and force pushed the other.  Looks like one or the other corrected the issue...

tmrobert8 commented:
Attached is a zip file containing the jar and the README.md for testing.  Discussion can be found over at [OpenHAB community](https://community.openhab.org/t/russound-rio/13731/8)

[org.openhab.binding.russound.zip](https://github.com/openhab/openhab2-addons/files/493020/org.openhab.binding.russound.zip)

tmrobert8 commented:
Attached is a new zip file with the following changes:

1. Added support for the undocumented source volume
2. Added support for the cover art image (in addition to url)
3. Fixed issues with source ipaddress (improperly documented in russound documentation)
4.  Handle bridge status changes better
5. Added state tracking to avoid resending same state over and over


[org.openhab.binding.russound.zip](https://github.com/openhab/openhab2-addons/files/506435/org.openhab.binding.russound.zip)

tmrobert8 commented:
Attached is the latest zip file containing both the jar and README.md.  There have been various bug fixes and I've added some new channels



[org.openhab.binding.russound.zip](https://github.com/openhab/openhab2-addons/files/516420/org.openhab.binding.russound.zip)

tmrobert8 commented:
Attached is the latest zip file containing both the jar and README.md. There have been various bug fixes and I've added some new channels

[org.openhab.binding.russound.zip](https://github.com/openhab/openhab2-addons/files/516752/org.openhab.binding.russound.zip)

Tomtibo commented:
Here is the new RIO API documentation.
I'm not sure if it's better to continu the discussion here, or on openhab forum...
I will give you news from my testing!
[RIO+Protocol+for+Third+Party+Integrators .pdf](https://github.com/openhab/openhab2-addons/files/520465/RIO.Protocol.for.Third.Party.Integrators.pdf)

kaikreuzer commented:
@tmrobert8 Before I go about reviewing this one - please let me know, if it is in a state to do so :-)

tmrobert8 commented:
Feel free to review - I have some minor changes and wanted to address the issue of the mca88x but a review will catch anything else (probably similar comment that you posted on the atlona one).

Note: I had to wipe my pc and reload (rebuilding the openhab dev environment now) so it may take me awhile to get around to fixing these things

tmrobert8 commented:
FYI - this should have conflicts now.  I was using the bridgeHandlerInitialized, which seems to have been removed (via smarthome#2087).  I'll need to update anyway to fix this - but welcome any other comments you have.

tmrobert8 commented:
Latest build

[org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/677732/org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip)

tmrobert8 commented:
Build specific to Tom:
[org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/678646/org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip)

kaikreuzer commented on addons/binding/org.openhab.binding.russound/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,324 @@
+        <description>Ethernet access point to Russound RIO control system (usually the main controller)</description>
+        
+        <channels>
+            <channel id="version" typeId="readonlystring"><label>Firmware Version</label><description>Firmware Version</description></channel>
```
firmware version should be added as a property, not as a channel

tmrobert8 replied:
Question for you (have never used properties before).  Can I set a property at any time?  All these 'channels' are information I get AFTER the plugin is initialized fully (and I start getting responses from the system)...

kaikreuzer replied:
Yes, anytime! Whenever you first get hold of the information. If the Thing is stored in a database (and not defined in a *.things file), it will even be persisted.

kaikreuzer commented on addons/binding/org.openhab.binding.russound/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,324 @@
+                <label>IP or Host Name</label>
+                <description>The IP or host name of the Russound RIO access point</description>
+            </parameter>
+            <parameter name="ping" type="integer" required="false">
```
mark as advanced

kaikreuzer commented on addons/binding/org.openhab.binding.russound/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,324 @@
+                <description>The ping interval in seconds to keep the connection alive</description>
+                <default>30</default>
+            </parameter>
+            <parameter name="retryPolling" type="integer" required="false">
```
mark as advanced

kaikreuzer commented on addons/binding/org.openhab.binding.russound/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,324 @@
+        <description>Controller of Zones, Sources, etc</description>
+
+        <channels>
+            <channel id="type" typeId="readonlystring"><label>Model Type</label><description>Model Type</description></channel>
```
should be a property, not a channel

kaikreuzer commented on addons/binding/org.openhab.binding.russound/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,324 @@
+        <channels>
+            <channel id="type" typeId="readonlystring"><label>Model Type</label><description>Model Type</description></channel>
+            <channel id="ipaddress" typeId="readonlystring"><label>IP Address</label><description>IP Address</description></channel>
+            <channel id="macaddress" typeId="readonlystring"><label>MAC Address</label><description>MAC Address</description></channel>
```
should be a property, not a channel

kaikreuzer commented on addons/binding/org.openhab.binding.russound/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,324 @@
+
+        <channels>
+            <channel id="type" typeId="readonlystring"><label>Model Type</label><description>Model Type</description></channel>
+            <channel id="ipaddress" typeId="readonlystring"><label>IP Address</label><description>IP Address</description></channel>
```
Hm, is this any different from the IP adress of the bridge? Anyhow, as this is not changing dynamically, it should merely be added as a property, if at all.

tmrobert8 replied:
Actually - I'm not sure about this one (that's why I left it in).  The API seems to indicate they can be (why else would they provide two different ip addresses).  However, in my system they are one and the same (one controller is they system interface and controls all other controllers).  However, the X-Series (which I do NOT own) is more decentralized and may be the reason for this.  Since I definitively don't know - I'd rather leave it (although I do agree that making it a property is best).

kaikreuzer commented on addons/binding/org.openhab.binding.russound/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,324 @@
+        </config-description>
+    </thing-type>
+
+    <channel-type id="string">
```
I strongly suggest to remove the types string, readonlystring and button.
Channel types should express a certain functionality - which can be very different for the different channels.
Reusing a type is really only meant for situations, where you have 1..n of the very same functionality, e.g. zones to control.

kaikreuzer commented on addons/binding/org.openhab.binding.russound/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,324 @@
+        <label>Russound Zone</label>
+        <description>Zone within a Controller</description>
+        
+        <channels>
```
make sure that only a few are regular channels and the majority are advanced channels.

kaikreuzer commented on addons/binding/org.openhab.binding.russound/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,324 @@
+        <label>Russound Source</label>
+        <description>Source (tuner, streamer, etc) within the Russound System</description>
+        
+        <channels>
```
make sure that only a few are regular channels and the majority are advanced channels.

kaikreuzer commented on addons/binding/org.openhab.binding.russound/README.md:
```diff
@@ -0,0 +1,441 @@
+
+The following configurations occur for each of the bridges/things:
+
+### Russound System
```
add an empty line after every headline

tmrobert8 commented:
hah - think this is huge - wait until you see the sony one (much bigger!).  I'll address the things you've mentioned (in addition to some of the stuff you mentioned in other ones) and recommit (doubtful until next week however).  Most of it's XML - so no big deal.  I've never worked with properties - but I'll look into it.  Not sure what you meant about the 'features.xml' (have no idea what that is).

kaikreuzer commented:
> Not sure what you meant about the 'features.xml' (have no idea what that is).

Very easy to explain: This file: https://github.com/openhab/openhab2-addons/blob/master/features/openhab-addons/src/main/feature/feature.xml
Adding your binding in there makes it being picked up by the distro - otherwise it will be missing!

tmrobert8 commented:
Ah - that would seem rather important then  :)

I'll add it (and do the same for the atlona and sony one in their commits)

tmrobert8 commented:
For Tom again...

[org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/679037/org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip)

tmrobert8 commented:
For Tom again

[org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/679046/org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip)

tmrobert8 commented:
For Tom again (#3):
[org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/679066/org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip)

tmrobert8 commented:
For Tom again (4):
[org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/679126/org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip)

tmrobert8 commented:
For Tom (5):

[org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/679157/org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip)

tmrobert8 commented:
Latest build - fixes MCA88 commands.  Please note that alot of changes have happened to the channels since the last posting (alot are now properties, many channel type changes).

[org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/679191/org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip)

tmrobert8 commented:
Kai,

I believe I got everything you mentioned in your review (and we solved the mca88x issue - very happy with that).  Feel free to do "review 2"

Tim

kaikreuzer commented:
You are much too fast for me! I won't find the time to review 6000+ LOC in this year anymore - but maybe next year ;-)

tmrobert8 commented on addons/binding/org.openhab.binding.russound/src/main/java/org/openhab/binding/russound/internal/net/BasicSocketSession.java:
```diff
@@ -0,0 +1,415 @@
+            throw new IllegalArgumentException("command cannot be null");
+        }
+
+        // if (command.trim().length() == 0) {
```
That whole class will be removed very shortly (I've been testing the SocketChannelSession implementation and it appears pretty stable).

tmrobert8 commented on addons/binding/org.openhab.binding.russound/README.md:
```diff
@@ -0,0 +1,461 @@
+# Russound Binding
+
+This binding provides integration with any Russound system that support the RIO protocol (all MCA systems, all X systems).  This binding provides  compatibility with RIO Protocol v1.7 (everything but the Media Managment functionality).  The protocol document can be found in the Russound Portal ("RIO Protocol for 3rd Party Integrators.pdf").  Please update to the latest firmware to provide full co ...
```
lol - Stupid Fingers like to Capitalize

tmrobert8 commented:
I'll make all these changes (need to look into the dispose stuff more - I need to send some commands to RIO on dispose and need to make sure those still happen).

Zone/Sources are definitely physical based (you can physically add more zones and more sources to the system).  I agree that favorites/banks/presets are 'murky' (especially favorites).  Why they are murky is that, while they are not physical, they are virtual in that you can add/remove them from the system (via russound's setup application).  Ex: I can create a bank #3 (without banks #1, #2 being defined) with presets (#2 and #4) on controller #2 then add bank#3.preset#4 as favorite #2 on zone #6 on controller #2.  When all was said and done, it seems easier to set them up as things because you can set openHAB in the exact same manner as you do in the application.  What's your thought on that?

tmrobert8 commented:
This discussion got me to go back and reread about [channel groups]((http://www.eclipse.org/smarthome/documentation/development/bindings/thing-definition.html) - which I have not used yet and I'm still a bit unfamilar with it.   On that page it states "A thing can only have direct channels or channel groups, but not both".  Is that true?  If so - that would be a show stopper as well (since a bank is assigned to a zone which has many other direct channels on it.  This would also seem to be a HUGE limitation to using channel groups (thinking of the atlona [hdbaset matrix switch] addon - the inputs/outputs could easily be modelled as channel groups (and would really simplify the addon) - but I'd need to mix those in with the other direct channels [power, etc]) and that statement seems to indicate it can't be mixed in.

kaikreuzer commented:
>  that statement seems to indicate it can't be mixed in.

No, I guess you misinterpret that. It is true that you have to decide for a Thing whether you want to use channel groups or not. If you use groups (because you need it for the banks), you would simply put all the "individual" channels into a "general" or "other" group.

kaikreuzer commented:
>  it seems easier to set them up as things because you can set openHAB in the exact same manner as you do in the application.

Question: Can the handler read all of this setup from the russound server (i.e. the number of banks and their ids) or does the user really have to replicate this setup manually in his Thing definitions and needs to keep them in sync manually?

tmrobert8 commented:
Currently, user needs to keep in sync.  I could infer it by API side effects (get "name" for controller[1].zone[1].bank[1] - no name, doesn't exist).  I'm pretty sure that's how the native application does it and it takes that application about 5 minutes to scan through my whole system.  

Eventually I wanted to put all this into a discovery process (port scan for the controller - then run through the API to discover what is or isn't setup).  That would take care of the initial setup - but if they modify an already setup system, there really is no way to detect any changes short of doing a full scan again (which is probably too intensive to do on a regular basis).

tmrobert8 commented:
This probably should be taken out of this discussion - but wanted to run this past you.  Let's say I change the bank into a channel group.  The zone would then have a channel-group "primary" which defines all the direct channels and then zero-to-many bank groups:
`
...
<channel-group id="primary" typeid="primary"/>
<channel-group id="bank1" typeid="bank"/>
<channel-group id="bank2" typeid="bank"/> 
...
<channel-group-type id="bank">
   <label>Russound Bank</label>
   ...
   <channels>...the generic bank channels (name, etc)</channels>
</channel-group-type>
`

A few questions:
1) How do we identify that "bank1" is russound bank id of 1 short of parsing the identifier from the channel's id?  With channel-type - I could identify it through channel config or possibly the tag - but channel-group-type doesn't seem to have those options?
2) How would I model presets (which would be channel groups within a bank)?  Doesn't appear that the channel-group-type allows embedded channel-group(s)?
3) I'd have no idea how to model favorites - which a attached to an individual preset but assigned to either the system or zone and can be changed via the API (added or removed)

kaikreuzer commented:
1. You will have to check it through the channel id, which will be "bank1#somechannel".
2. Correct, there is only one level, you cannot put groups into groups. You would add a simple "favorite" channel to your bank group (as a dynamic channel, which isn't define in the group type).
3. Having a closer look at it, I would suggest to heavily reduce the number of channels. It seems as if you have defined a channel for every possible command. Instead, I would create a single "favorite" channel that has StringType and which provides the list of available favorite names as options. Sending a string would then do a "restore" of the favorite. I would leave save&delete out of scope, this is probably rather something the user would do through the dedicated app, not necessarily through openHAB.

tmrobert8 commented:
Kai,

The problem is - I use all that functionality (that's why I wrote it in).  Example: presets and favorites are modified based on room presence (ie my wife's presets/favorites follow her around, as do mine and my sons).

If you don't mind, I'm going to leave this one alone - I really think the way I have it modeled is the way the RIO system is put together and it feel less like 'forcing' the channel grouping onto it.

I will however be modifying the atlona binding to use channel groupings - the input/outputs are much better modeled with channel groups rather than how I have it now.

Tim

tmrobert8 commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -200,9 +200,13 @@
+
     <feature name="openhab-binding-samsungtv" description="Samsung TV Binding" version="${project.version}">
         <feature>openhab-runtime-base</feature>
-        <feature>openhab-transport-upnp</feature>
```
Ack!  Meant to remove that line from the russound one!  I'll revert that one...

kaikreuzer commented:
> presets and favorites are modified based on room presence

So you are never together with your wife in the same room? ;-)

Well, ok, I don't have a Russound, so I cannot really assess how it is all set up and used, so let's keep your implemented modelling. 
Nonetheless, you might think about combining the save/delete/restore channels into one "command" channel that takes predefined command (using "OFF" for "zone favorite" is really ugly...). After all, for none of these channels, a state makes any sense, so there is no reason for having them as separate channels.

tmrobert8 commented:
Kai - just committed all the changes that should address these notes.  BTW - excellent suggestion on the command channels - makes alot more sense that way (not sure why I didn't think about it).

tmrobert8 commented:
Latest build - fixes MCA88 commands. Please note that alot of changes have happened to the channels since the last posting (alot are now properties, many channel type changes).

Latest build 2 - changed from a basic socket implementation to a socket channel implementation and fixed issue reconnecting if network drops

Latest build 3 - fixed issue with warning message during disconnnect, fixed issue when reloading configuration.

Latest build 4 - implemented many of the changes from the review.

[org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/684891/org.openhab.binding.russound-2.0.0-SNAPSHOT.jar.zip)

kaikreuzer commented on addons/binding/org.openhab.binding.russound/META-INF/MANIFEST.MF:
```diff
@@ -20,8 +20,8 @@ Import-Package:
 Service-Component: OSGI-INF/*.xml
 Export-Package: org.openhab.binding.russound,
- org.openhab.binding.russound.rio
+ org.openhab.binding.russound.internal.rio
```
no, you should not export it anymore now :-)

tmrobert8 commented:
Done - you may want to hold off on the actual merge until Tom tests using his mca-88x.  I'll post a message when he seems good with it (modifying the atlona addon right now with all the comments and changes - don't bother reviewing that one)

Tomtibo commented:
Hi! I've tested the latest snapshot and all is working fine with the mca-88x! I've not tested favorites and banks, I don't know how or if they works with the mca-88x. Nice work guys!!

tmrobert8 commented:
Kai,

Feel free to merge then!

Tim

kaikreuzer commented:
Thanks again!
